File: App/General/Layout/MBStateMachineView.m

#import "MBStateMachineView.h"
#import <RFKit/UIView+RFKit.h>

@interface MBStateMachineView ()
@property (nonatomic, strong) NSHashTable<id<MBStateMachineViewDelegate>> *delegates;
@property (nonatomic, weak) UIView *contentView;
@end

@implementation MBStateMachineView
RFInitializingRootForUIView

- (void)onInit {
    // Initialization code
}

- (void)afterInit {
}

- (void)awakeFromNib {
    [super awakeFromNib];
    if (!self.viewSource) {
        for (UIView<MBStateMachineViewIdentifying> *subView in self.subviews) {
            if (![subView respondsToSelector:@selector(stateIdentifier)]) continue;

            if ([self.state isEqualToString:subView.stateIdentifier]) {
                subView.hidden = NO;
            }
            else {
                subView.hidden = YES;
            }
        }
    }
}

- (NSHashTable *)delegates {
    if (_delegates) {
        _delegates = [NSHashTable weakObjectsHashTable];
    }
    return _delegates;
}

- (void)addDelegate:(nullable id<MBStateMachineViewDelegate>)aDelegate {
    [self.delegates addObject:aDelegate];
}

- (void)removeDelegate:(id<MBStateMachineViewDelegate>)aDelegate {
    [self.delegates removeObject:aDelegate];
}

- (void)setState:(NSString *)state {
    [self stateChangedFromState:_state toState:state];
    _state = state;
}

- (UIView<MBStateMachineViewIdentifying> *)viewForState:(NSString *)state {
    id viewSource = self.viewSource;
    if (viewSource) {
        NSString *newViewSelector = [NSString stringWithFormat:@"viewFor%@", state];
        id view = [viewSource valueForKey:newViewSelector];
        NSAssert(!view || [view isKindOfClass:[UIView class]], @"%@ not a UIView", newViewSelector);
        return view;
    }

    for (UIView<MBStateMachineViewIdentifying> *subView in self.subviews) {
        if ([subView respondsToSelector:@selector(stateIdentifier)]
            && [state isEqualToString:subView.stateIdentifier]) {
            return subView;
        }
    }
    return nil;
}

- (void)stateChangedFromState:(nullable NSString *)oldStats toState:(nullable NSString *)newState {
    id viewSource = self.viewSource;

    if (![oldStats isEqualToString:newState]
        && newState) {
        UIView *oldView = self.contentView;
        UIView<MBStateMachineViewIdentifying> *newView = [self viewForState:newState];

        if (oldView != newView) {
            if (viewSource) {
                [oldView removeFromSuperview];
                [self addSubview:newView resizeOption:RFViewResizeOptionFill];
                newView.autoresizingMask = UIViewAutoresizingFlexibleSize;
            }
            else {
                oldView.hidden = YES;
                if ([newView respondsToSelector:@selector(dontResizeWhenStateChanged)]
                    && newView.dontResizeWhenStateChanged) {

                }
                else {
                    newView.frame = self.bounds;
                }
                newView.hidden = NO;
            }
            self.contentView = newView;
        }
    }
    self.hidden = !newState;

    for (id<MBStateMachineViewDelegate> delegate in self.delegates) {
        if ([delegate respondsToSelector:@selector(stateMachineView:didChangedStateFromState:toState:)]) {
            [delegate stateMachineView:self didChangedStateFromState:oldStats toState:newState];
        }
    }
}

@end


@implementation MBStateMachineSubview

@end

File: App/General/Layout/MBStateMachineView.swift
/*
 MBStateMachineView

 Copyright © 2018, 2023 BB9z.
 Copyright © 2015 Beijing ZhiYun ZhiYuan Technology Co., Ltd.
 https://github.com/BB9z/iOS-Project-Template

 The MIT License
 https://opensource.org/licenses/MIT
 */

import B9MulticastDelegate
import UIKit

// @MBDependency:2
/**
 划定一个区域，改变一个标识符，这块区域将显示不同的内容，这就是状态机 view 的作用

 frame 现在是固定的
 */
class MBStateMachineView: UIView {
    weak var viewSource: AnyObject?
    weak var contentView: UIView?
    weak var delegate: MBStateMachineViewDelegate?
    var state: String? {
        didSet {
            stateChanged(from: oldValue, to: state)
        }
    }

    private let delegates = MulticastDelegate<MBStateMachineViewDelegate>()

    override func awakeFromNib() {
        super.awakeFromNib()
        if viewSource != nil { return }
        for subView in subviews {
            if let identifyingView = subView as? MBStateMachineViewIdentifying,
               let stateIdentifier = identifyingView.stateIdentifier {
                identifyingView.isHidden = state != stateIdentifier
            }
        }
    }

    private func view(for state: String) -> UIView? {
        if let source = viewSource {
            let key = "viewFor\(state)"
            if let result = source.value?(forKey: key) as? UIView {
                return result
            }
        }
        return subviews.first {
            ($0 as? MBStateMachineViewIdentifying)?.stateIdentifier == state
        }
    }

    private func stateChanged(from oldState: String?, to newState: String?) {
        if let newState = newState,
           oldState != newState {
            let oldView = contentView
            let newView = view(for: newState)

            if oldView != newView {
                if viewSource != nil {
                    oldView?.removeFromSuperview()
                    if let view = newView {
                        addSubview(view, resizeOption: .fill)
                        view.autoresizingMask = [.flexibleWidth, .flexibleHeight]
                    }
                } else {
                    oldView?.isHidden = true
                    if let view = newView as? MBStateMachineViewIdentifying,
                       view.dontResizeWhenStateChanged {
                        // noop
                    } else {
                        newView?.frame = bounds
                    }
                    newView?.isHidden = false
                }
                contentView = newView
            }
        }
        isHidden = newState == nil
        delegates.invoke {
            $0.stateMachineView(self, didChangedStateFromState: oldState, toState: newState)
        }
    }
}

protocol MBStateMachineViewDelegate: AnyObject {
    func stateMachineView(_ view: MBStateMachineView, didChangedStateFromState oldStats: String?, toState newState: String?)
}

protocol MBStateMachineViewIdentifying: UIView {
    var stateIdentifier: String? { get }
    var dontResizeWhenStateChanged: Bool { get }
}

class MBStateMachineSubview: UIView, MBStateMachineViewIdentifying {
    @IBInspectable var stateIdentifier: String?
    @IBInspectable var dontResizeWhenStateChanged: Bool = false
}
